src/radiation_system.hpp

```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+			Frad_t0[1] = consPrev(i, j, k, x2RadFlux_index + numRadVars_ * g);
+			Frad_t0[2] = consPrev(i, j, k, x3RadFlux_index + numRadVars_ * g);
 
-		// 3. Compute conservative gas momentum update
-		amrex::GpuArray<amrex::Real, 3> dF{};
-		amrex::GpuArray<amrex::Real, 3> dMomentum{};
+			for (int n = 0; n < 3; ++n) {
+				AMREX_ASSERT((advectionFluxes(i, j, k, n) == 0.0)); // CCH: for sanity check
+				Frad_t1[n] = (Frad_t0[n] + (dt * advectionFluxes(i, j, k, n))) / (1.0 + rho * kappaFVec[g] * chat * dt);
+				// Compute conservative gas momentum update
+				dMomentum[n] = -(Frad_t1[n] - Frad_t0[n]) / (c * chat);
+			}
 
-		for (int n = 0; n < 3; ++n) {
-			dF[n] = Frad_t1[n] - Frad_t0[n];
-			dMomentum[n] = -dF[n] / (c * chat);
-		}
+			// update radiation flux on consNew at current group
+			consNew(i, j, k, x1RadFlux_index + numRadVars_ * g) = Frad_t1[0];
+			consNew(i, j, k, x2RadFlux_index + numRadVars_ * g) = Frad_t1[1];
+			consNew(i, j, k, x3RadFlux_index + numRadVars_ * g) = Frad_t1[2];
 
-		consNew(i, j, k, x1GasMomentum_index) += dMomentum[0];
-		consNew(i, j, k, x2GasMomentum_index) += dMomentum[1];
-		consNew(i, j, k, x3GasMomentum_index) += dMomentum[2];
+			// compute the gas momentum after adding the current group's momentum
+			consNew(i, j, k, x1GasMomentum_index) += dMomentum[0];
+			consNew(i, j, k, x2GasMomentum_index) += dMomentum[1];
+			consNew(i, j, k, x3GasMomentum_index) += dMomentum[2];
+		}
 
+		// update gas momentum from radiation momentum of the current group
 		if constexpr (gamma_ != 1.0) {
 			amrex::Real x1GasMom1 = consNew(i, j, k, x1GasMomentum_index);
 			amrex::Real x2GasMom1 = consNew(i, j, k, x2GasMomentum_index);
 			amrex::Real x3GasMom1 = consNew(i, j, k, x3GasMomentum_index);
 
 			// 4a. Compute radiation work terms
 			amrex::Real const Egastot1 = ComputeEgasFromEint(rho, x1GasMom1, x2GasMom1, x3GasMom1, Egas_guess);
-			amrex::Real dErad_work = NAN;
+			amrex::Real dErad_work = 0.;
 
 			if constexpr (compute_v_over_c_terms_) {
-				// compute difference in gas kinetic energy before and after momentum
-				// update
+				// compute difference in gas kinetic energy before and after momentum update
 				amrex::Real const Ekin1 = Egastot1 - Egas_guess;
 				amrex::Real const dEkin_work = Ekin1 - Ekin0;
 				// compute loss of radiation energy to gas kinetic energy
 				dErad_work = -(c_hat_ / c_light_) * dEkin_work;
+
+				// compute energyLossFractions
+				quokka::valarray<double, nGroups_> energyLossFractions{};
+				for (int g = 0; g < nGroups_; ++g) {
+					Frad_t1[0] = consNew(i, j, k, x1RadFlux_index + numRadVars_ * g);
+					Frad_t1[1] = consNew(i, j, k, x2RadFlux_index + numRadVars_ * g);
+					Frad_t1[2] = consNew(i, j, k, x3RadFlux_index + numRadVars_ * g);
+					energyLossFractions[g] = kappaFVec[g] * (x1GasMom1 * Frad_t1[0] + x2GasMom1 * Frad_t1[1] + x3GasMom1 * Frad_t1[2]);
+				}
+				// TODO(CCH): tricky here. Have to avoid negavie radiation energy.
```

Comment:
Can you explain the negative energy problem you encountered?

Reply:
Never mind this. It shouldn't be a TODO and I removed it.